superduperdb/vector_search/atlas.py

```diff
@@ -0,0 +1,198 @@
+
+        indexing_key = vi.indexing_listener.key
+        if indexing_key.startswith('_outputs'):
+            indexing_key = indexing_key.split('.')[1]
+        assert isinstance(vi.indexing_listener.model, Model)
+        assert isinstance(collection, str), 'Collection is required to be a string'
+        indexing_model = vi.indexing_listener.model.identifier
+        indexing_version = vi.indexing_listener.model.version
+        output_path = f'_outputs.{indexing_key}.{indexing_model}.{indexing_version}'
+
+        return MongoAtlasVectorSearcher(
+            identifier=vi.identifier,
+            dimensions=vi.dimensions,
+            measure=vi.measure,
+            output_path=output_path,
+            collection=collection,
+        )
+
+    def _replace_document_with_vector(self, step):
+        step = copy.deepcopy(step)
+        assert "like" in step['$vectorSearch']
+        vector = step['$vectorSearch']['like']
+        step['$vectorSearch']['queryVector'] = vector
+
+        step['$vectorSearch']['path'] = self.output_path
+        step['$vectorSearch']['index'] = self.identifier
+        del step['$vectorSearch']['like']
+        return step
+
+    def _prepare_pipeline(self, pipeline):
+        pipeline = copy.deepcopy(pipeline)
+        try:
+            search_step = next(
+                (i, step) for i, step in enumerate(pipeline) if '$vectorSearch' in step
+            )
+        except StopIteration:
+            return pipeline
+        pipeline[search_step[0]] = self._replace_document_with_vector(
+            search_step[1],
+        )
+        return pipeline
+
+    def _find(self, h, n=100):
+        h = self.to_list(h)
+        pl = [
+            {
+                "$vectorSearch": {
+                    'like': h,
+                    "limit": n,
+                    'numCandidates': n,
+                }
+            },
+            {'$addFields': {'score': {'$meta': 'vectorSearchScore'}}},
+        ]
+        pl = self._prepare_pipeline(
+            pl,
+        )
+        cursor = self.index.aggregate(pl)
+        scores = []
+        ids = []
+        for vector in cursor:
+            scores.append(vector['score'])
+            ids.append(str(vector['_id']))
+        return ids, scores
+
+    def find_nearest_from_id(self, id: str, n=100, within_ids=None):
+        h = self.index.find_one({'id': id})
+        return self.find_nearest_from_array(h, n=n, within_ids=within_ids)
+
+    def find_nearest_from_array(self, h, n=100, within_ids=None):
+        return self._find(h, n=n)
+
+    def add(self, items):
+        items = list(map(lambda x: x.to_dict(), items))
+        if not CFG.cluster.vector_search == CFG.data_backend:
+            self.index.insert_many(items)
+
+    def delete(self, items):
+        ids = list(map(lambda x: x.id, items))
+        if not CFG.cluster.vector_search == CFG.data_backend:
+            self.index.delete_many({'id': {'$in': ids}})
```

Comment:
Do we need to delete the data if `CFG.cluster.vector_search == CFG.data_backend`?

Reply:
No, since that is handled by `db.execute(<delete_query>)`

Reply:
ok, got it